main.go

```diff
@@ -1,21 +1,279 @@
+	gParams.HistoryLength = 6
+	gParams.HistoryGossip = 3
+	return gParams
+}
+
+// pubsubOptions creates a list of options to configure our router with.
+func pubsubOptions() []pubsub.Option {
+	psOpts := []pubsub.Option{
+		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign),
+		pubsub.WithNoAuthor(),
+		pubsub.WithPeerOutboundQueueSize(600),
+		pubsub.WithMaxMessageSize(10 * 1 << 20),
+		pubsub.WithValidateQueueSize(600),
+		pubsub.WithGossipSubParams(pubsubGossipParam()),
+	}
+
+	return psOpts
+}
+
+func convToPrivKey(input string) (crypto.PrivKey, error) {
+	hash := sha256.New()
+	hash.Write([]byte(input))
+	seed := hash.Sum(nil)
+
+	priv := ed25519.NewKeyFromSeed(seed)
+	privKey, err := crypto.UnmarshalEd25519PrivateKey(priv)
+	if err != nil {
+		return nil, err
+	}
+
+	return privKey, nil
+}
+
+func getPeerID(input string) peer.ID {
+	privKey, err := convToPrivKey(input)
+	if err != nil {
+		panic(err)
+	}
+
+	// Convert the hash to a peer.ID
+	peerID, err := peer.IDFromPrivateKey(privKey)
+	if err != nil {
+		panic(err)
+	}
+
+	return peerID
+}
+
+func publish(ctx context.Context, t *pubsub.Topic, data []byte, opts ...pubsub.PubOpt) error {
+	// Wait for at least 1 peer to be available to receive the published message.
+	for {
+		if len(t.ListPeers()) > 0 {
+			return t.Publish(ctx, data, opts...)
+		}
+
+		select {
+		case <-ctx.Done():
+			return errors.New("unable to find requisite number of peers, 0 peers found to publish to")
+		default:
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+}
+
+func leave(t *pubsub.Topic) error {
+	if err := t.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func connectToRandomPeers(ctx context.Context, h host.Host, num int) int {
+	rand.Seed(time.Now().UnixNano())
+
+	var connected atomic.Uint64
+	var wg sync.WaitGroup
+	for i := 0; i < num; i++ {
+		hname := "node" + strconv.Itoa(rand.Intn(*countFlag)+1)
+		ip, err := resolveHostname(hname)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+		fmt.Printf("Resolved %s to %s\n", hname, ip)
+
+		maddr := "/ip4/" + ip + "/tcp/" + PORT + "/p2p/" + getPeerID(hname).String()
+		info, err := peer.AddrInfoFromString(maddr)
+		if err != nil {
+			fmt.Println(err)
+			continue
+		}
+		fmt.Printf("Connecting to %s\n", maddr)
+
+		// make each dial non-blocking
+		wg.Add(1)
+		go func(pi peer.AddrInfo) {
+			defer wg.Done()
+			ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+			defer cancel()
+			if err := h.Connect(ctx, pi); err != nil {
+				fmt.Println("couldn't dial ", pi.Addrs)
+			}
+			connected.Add(1)
+		}(*info)
+	}
+	wg.Wait()
+	return int(connected.Load())
+}
+
 func main() {
+	fmt.Println("Starting a node")
+
 	flag.Parse()
 	hostname, err := os.Hostname()
 	if err != nil {
 		panic(err)
 	}
 	fmt.Printf("Hostname: %s\n", hostname)
 	fmt.Printf("Count: %d\n", *countFlag)
+	fmt.Printf("Degree: %d\n", *degreeFlag)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	transports := libp2p.ChainOptions(
```

Comment:
TCP is already specifiec in `listenAddrs`. I think there is no need to do it again here.